File: win/Sources/OverlayCoordinator.cs
using System;
using System.Collections.Generic;
using System.Windows;
using System.Windows.Media;
using System.Windows.Media.Imaging;

namespace GlassFace;

/// <summary>
/// Owns every display's overlay and renders camera frames into them. As the camera's
/// <see cref="IFrameConsumer"/>, it copies each frame into one shared
/// <see cref="WriteableBitmap"/> that all the per-screen images display, and applies opacity
/// across all of them. It renders; it does not manage the camera session lifecycle
/// (that's the coordinator's job).
/// </summary>
public sealed class OverlayCoordinator : IFrameConsumer
{
    public float Opacity { get; private set; }

    private readonly List<ScreenOverlay> _overlays = new();
    private WriteableBitmap? _preview;

    public OverlayCoordinator(AppConfiguration config)
    {
        Opacity = config.DefaultOpacity;
    }

    /// <summary>Creates one overlay window per connected display and shows them.</summary>
    public void BuildOverlays()
    {
        foreach (var bounds in NativeMethods.EnumerateMonitors())
            _overlays.Add(new ScreenOverlay(bounds, Opacity));
        foreach (var overlay in _overlays)
            overlay.Present();
    }

    /// <summary>Clamps, applies, and returns the resulting opacity.</summary>
    public float SetOpacity(float value)
    {
        Opacity = Math.Clamp(value, 0f, 1f);
        foreach (var overlay in _overlays)
            overlay.SetOpacity(Opacity);
        return Opacity;
    }

    // MARK: IFrameConsumer — called off the UI thread by CameraService.

    public void Consume(CameraFrame frame)
    {
        var dispatcher = Application.Current?.Dispatcher;
        if (dispatcher is null || dispatcher.HasShutdownStarted) return;

        // Synchronous so the camera's reusable buffer is safe to overwrite once we return.
        try { dispatcher.Invoke(() => Render(frame)); }
        catch (System.Threading.Tasks.TaskCanceledException) { /* shutting down */ }
    }

    private void Render(CameraFrame frame)
    {
        if (_preview is null || _preview.PixelWidth != frame.Width || _preview.PixelHeight != frame.Height)
        {
            _preview = new WriteableBitmap(frame.Width, frame.Height, 96, 96, PixelFormats.Bgr32, null);
            foreach (var overlay in _overlays)
                overlay.SetSource(_preview);
        }

        _preview.WritePixels(new Int32Rect(0, 0, frame.Width, frame.Height),
            frame.Bgra, frame.Width * 4, 0);
    }
}

File: win/Sources/OverlayCoordinator.cpp
#include "pch.h"
#include "OverlayCoordinator.h"
#include "NativeHelpers.h"
#include <algorithm>

using namespace winrt;
using namespace winrt::Windows::Graphics::Imaging;
using namespace winrt::Microsoft::UI::Dispatching;
using namespace winrt::Microsoft::UI::Xaml::Media::Imaging;

namespace glassface
{
    OverlayCoordinator::OverlayCoordinator(AppConfiguration const& config)
        : m_opacity(config.defaultOpacity)
    {
    }

    void OverlayCoordinator::BuildOverlays()
    {
        m_dispatcher = DispatcherQueue::GetForCurrentThread();
        m_source = SoftwareBitmapSource();

        for (auto const& bounds : EnumerateMonitors())
            m_overlays.push_back(std::make_unique<ScreenOverlay>(bounds, m_opacity, m_source));

        for (auto const& overlay : m_overlays)
            overlay->Present();
    }

    float OverlayCoordinator::SetOpacity(float value)
    {
        m_opacity = std::clamp(value, 0.0f, 1.0f);
        for (auto const& overlay : m_overlays)
            overlay->SetOpacity(m_opacity);
        return m_opacity;
    }

    void OverlayCoordinator::Consume(SoftwareBitmap const& bitmap)
    {
        if (!m_dispatcher) return;

        // Drop frames while one is still being uploaded, so we never queue up a backlog.
        bool expected = false;
        if (!m_busy.compare_exchange_strong(expected, true)) return;

        SoftwareBitmap frame = bitmap;
        bool queued = m_dispatcher.TryEnqueue([this, frame]() { RenderOnUi(frame); });
        if (!queued) m_busy.store(false);
    }

    fire_and_forget OverlayCoordinator::RenderOnUi(SoftwareBitmap bitmap)
    {
        // Runs on the UI thread (enqueued via the dispatcher); SetBitmapAsync resumes here too.
        try { co_await m_source.SetBitmapAsync(bitmap); }
        catch (...) {}
        m_busy.store(false);
    }
}
